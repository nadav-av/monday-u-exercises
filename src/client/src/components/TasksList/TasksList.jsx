import React, { useState, useEffect, useCallback } from "react";
import PropTypes from "prop-types";
import TaskItem from "../TaskItem/TaskItemConnector";
import { DragDropContext, Draggable, Droppable } from "react-beautiful-dnd";
import { COMPLETED, UNCOMPLETED } from "../../services/globalConsts";
import { useDebounce } from "../../hooks/useDebounce";
import { DEBOUNCE_RATE_MS } from "../../services/globalConsts";
import "./taskList.css";

const TasksList = ({
  tasks,
  setTasksAction,
  searchInput,
  statusFilter,
  setPresentedTasksNum,
}) => {
  const setTasks = useCallback(
    (tasks) => {
      setTasksAction(tasks);
    },
    [setTasksAction]
  );

  const [filteredTasks, setFilteredTasks] = useState([]);

  const debouncedSearchInput = useDebounce(searchInput, DEBOUNCE_RATE_MS);

  useEffect(() => {
    const searchFilteredTasks = tasks.filter((task) => {
      return task.itemName
        .toLowerCase()
        .includes(debouncedSearchInput.toLowerCase());
    });

    let statusFilteredTasks = [];

    switch (statusFilter) {
      case COMPLETED:
        statusFilteredTasks = searchFilteredTasks.filter(
          (task) => task.status === true
        );
        break;
      case UNCOMPLETED:
        statusFilteredTasks = searchFilteredTasks.filter(
          (task) => task.status === false
        );
        break;
      default:
        statusFilteredTasks = searchFilteredTasks;
        break;
    }
    const positionSortedTasks = statusFilteredTasks.sort((a, b) => {
      return a.position - b.position;
    });
    setFilteredTasks(positionSortedTasks);
    setPresentedTasksNum(positionSortedTasks.length);
  }, [tasks, debouncedSearchInput, statusFilter]);

  const handleOnDragEnd = async (result) => {
    const items = [...tasks];
    const [removed] = items.splice(result.source.index, 1);
    items.splice(result.destination.index, 0, removed);
    setTasks(items);
  };

  return (
    <div data-testid="test-task-list">
      <DragDropContext onDragEnd={handleOnDragEnd}>
        <Droppable droppableId="tasks">
          {(provided) => (
            <ul
              className="tasks"
              {...provided.droppableProps}
              ref={provided.innerRef}
            >
              {filteredTasks.map((task, index) => (
                <Draggable
                  key={task.id.toString()}
                  draggableId={task.id.toString()}
                  index={index}
                >
                  {(provided) => (
                    <li
                      className="tasks-list"
                      {...provided.draggableProps}
                      {...provided.dragHandleProps}
                      ref={provided.innerRef}
                    >
                      <TaskItem task={task} />
                    </li>
                  )}
                </Draggable>
              ))}
              {provided.placeholder}
            </ul>
          )}
        </Droppable>
      </DragDropContext>
    </div>
  );
};

TasksList.propTypes = {
  setPresentedTasksNum: PropTypes.func.isRequired,
};

export default TasksList;
