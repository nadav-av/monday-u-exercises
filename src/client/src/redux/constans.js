const actions = {
  ADD_TASK: "ADD_TASK",
  REMOVE_TASK: "REMOVE_TASK",
  UPDATE_TASK: "UPDATE_TASK",
  REMOVE_ALL_TASKS: "REMOVE_ALL_TASKS",
  SET_TASKS: "SET_TASKS",
};

export default actions;
