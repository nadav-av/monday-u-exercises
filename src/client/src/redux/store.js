import { configureStore } from "@reduxjs/toolkit";
import thunkMiddleware from "redux-thunk";
import tasksReducer from "./reducers/taskReducer";
import actionBarReducer from "./reducers/actionBarReducer";
import errorHandleReducer from "./reducers/errorHandleReducer";
import logger from "redux-logger";

export const store = configureStore({
  reducer: {
    tasks: tasksReducer,
    actionBar: actionBarReducer,
    errorHandle: errorHandleReducer,
  },
  middleware: [thunkMiddleware, logger],
  preloadedState: {},
});
