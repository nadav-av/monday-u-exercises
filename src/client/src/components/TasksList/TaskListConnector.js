import { connect } from "react-redux";
import { getTasks } from "../../redux/selectors/tasksSelector";
import {
  removeTaskAction,
  setTasksAction,
  updateTaskAction,
} from "./../../redux/actions/tasks_actions";
import { bindActionCreators } from "redux";
import TasksList from "./TasksList";

const mapStateToProps = (state, ownProps) => {
  const tasks = getTasks(state);
  return { tasks };
};

const mapDispatchToProps = (dispatch, ownProps) => {
  return bindActionCreators(
    { updateTaskAction, setTasksAction, removeTaskAction },
    dispatch
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(TasksList);
