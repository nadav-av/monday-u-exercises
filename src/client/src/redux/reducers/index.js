import tasks from "./taskReducer";
import { combineReducers } from "redux";

const allReducers = combineReducers({
  tasks,
});

export default allReducers;
