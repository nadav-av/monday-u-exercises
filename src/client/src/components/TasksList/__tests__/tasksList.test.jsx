import renderer from "react-test-renderer";
import TasksList from "./../TaskListConnector";
import { store } from "./../../../redux/store";
import { Provider } from "react-redux";

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <TasksList />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
