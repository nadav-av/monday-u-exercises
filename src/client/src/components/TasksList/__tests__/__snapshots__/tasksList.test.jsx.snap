// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  data-testid="test-task-list"
>
  <ul
    className="tasks"
    data-rbd-droppable-context-id="0"
    data-rbd-droppable-id="tasks"
  />
</div>
`;
